### PR TITLE
[BD-21] Make use of edx_toggles newest API for testing

### PR DESCRIPTION
**Pull request by regisb**
**Description:** Make use of edx_toggles API for overriding waffle switches in tests

**JIRA:** https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943

**Dependencies:** https://github.com/edx/edx-toggles/pull/75 :heavy_check_mark: merged

**Reviewers:**
- [x] @robrap

**Merge checklist:**
- [x] All reviewers approved
- [x]  @robrap to confirm `deprecated_waffle_utils` custom attribute data exists before landing this
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @regisb! I've created [BLENDED-642](https://openedx.atlassian.net/browse/BLENDED-642) to keep track of it in Jira.   More details are on the [BD-21](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation) project page.


When this pull request is ready, tag your edX technical lead.

**Comment by regisb** [comment]
This is ready for review @robrap.

File: completion/test_utils.py
Comment (regisb) [reply completion/test_utils.py:69]: This is unfortunately not working, as we remove the `.override` methods from the waffle namespace classes in edx-platform. edx-platform tests themselves depend on edx-completion's test_utils, so tests will not pass until we merge this.

Comment (robrap) [reply completion/test_utils.py:69]: 1. I presume you mean in this PR: https://github.com/edx/edx-platform/pull/25417/files?
2. In the edx-platform, I'd rather deprecate and then remove.  I usually add something like `set_custom_attribute('deprecated_THING_TO_BE_DEPRECATED', True)` to the deprecated methods so I can track that there really is no usage before removal.

We can still land this, but it might be better to land it after, rather than before, so I can see that the `deprecated_X` custom attribute was firing and finally stops. The main point is to see if anything else is using it that we didn't notice before removal.

File: completion/waffle.py
Comment (robrap) [inline completion/waffle.py:35]: It doesn't make sense to me to have a general function, but annotations for a single switch.
Could this instead be?:
```
# .. toggle_name: completion.enable_completion_tracking
# .. toggle_implementation: WaffleSwitch
# .. toggle_default: False
# .. toggle_description: Indicates whether or not to track completion of individual blocks. Keeping this disabled
#     will prevent creation of BlockCompletion objects in the database, as well as preventing completion-related
#     network access by certain xblocks.
# .. toggle_use_cases: open_edx
ENABLE_COMPLETION_TRACKING_SWITCH = WaffleSwitch(waffle(), ENABLE_COMPLETION_TRACKING, module_name=__name__)
```

File: completion/test_utils.py
Comment (regisb) [reply completion/test_utils.py:69]: Yes, this is linked to https://github.com/edx/edx-platform/pull/25417.

The override functions are test methods, so my thinking was that it would be overkill to go through the deprecation process for them. But sure, I can do that.

> We can still land this, but it might be better to land it after, rather than before, so I can see that the deprecated_X custom attribute was firing and finally stops. The main point is to see if anything else is using it that we didn't notice before removal.

This needs to be merged before https://github.com/edx/edx-platform/pull/25417, or unit tests will fail in edx-platform -- unless we go through a deprecation process.

File: completion/waffle.py
Comment (regisb) [reply completion/waffle.py:35]: Yes, this makes much more sense.

**Comment by regisb** [comment]
@robrap I updated this PR -- and did not squash my changes this time :)

File: completion/test_utils.py
Comment (robrap) [inline completion/test_utils.py:72]: Did you mean to delete this whole method?  It seems like you replaced it with setup/teardown.

**Review by robrap — COMMENTED** [review]
In terms of timing, assuming we add a custom attribute for monitoring deprecation of the edx-platform `WaffleSwitch` and `WaffleSwitchNamespace` before removing them, I'd like to land this after the edx-platform deprecation, and before the removal.

File: completion/test_utils.py
Comment (regisb) [reply completion/test_utils.py:72]: No, the completion mixin has two mechanisms to override the waffle switch:

1. The first is the setUpClass/tearDownClass trick by which the override switch default value is set to the class' COMPLETION_SWITCH_ENABLED attribute.
2. The second is this decorator, which overrides the default value set by the setup/tearDown hack.

**Comment by regisb** [comment]
> I'd like to land this after the edx-platform deprecation, and before the removal.

Sounds good. I'll add back the override method to https://github.com/edx/edx-platform/pull/25417 so that we can merge this completion PR after.

File: completion/test_utils.py
Comment (robrap) [reply completion/test_utils.py:72]: I somewhat see. :) I find it confusing that we have two functions named `override_waffle_switch`, and I'm not clear on where the class level one is called.  Could you rename the class level `override_waffle_switch` to `override_completion_tracking_switch`?

Comment (regisb) [reply completion/test_utils.py:72]: There is just one `override_waffle_switch`, which is the one from edx-toggles . The other one is called `override_completion_switch`: https://github.com/edx/completion/pull/118/files/67f276a0231e80ebbb6009151cf50d0d3edf381a#diff-f47a4a90118f43738723e016b4338cce98e746a466d978c89127abab6c66b5eaR114

The setUpClass/tearDownClass methods are automatically called by unittest on test class creation: https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.setUpClass

Comment (robrap) [reply completion/test_utils.py:72]: That's not what I see in this file: https://github.com/edx/completion/blob/67f276a0231e80ebbb6009151cf50d0d3edf381a/completion/test_utils.py

Comment (regisb) [reply completion/test_utils.py:72]: Right! So that makes three overriding mechanisms...

Comment (regisb) [reply completion/test_utils.py:72]: Unfortunately, I cannot rename `CompletionWaffleTestMixin.override_waffle_switch` because this method is used in many places in edx-platform... :sweat:

Comment (robrap) [reply completion/test_utils.py:72]: Got it.  Options include:
1. Ignore and do nothing.
2. Adjust the import to provide a second name.
3. Introduce a new name for the one in the class, and keep the older version and mark it as deprecated.  We could then move to the newer name at some point in the future.
I'll leave this to you to decide.

Comment (regisb) [reply completion/test_utils.py:72]: I'm going to go with option 1...

File: completion/waffle.py
Comment (robrap) [inline completion/waffle.py:30]: @regisb: Do you know that this is meant to be a permanent flag, or was this just a guess?

**Review by robrap — APPROVED** [review]
I plan on delaying merge until I can see the monitoring custom attribute from the edx-platform PR.  Please ensure this is squashed as you wish.

File: completion/waffle.py
Comment (regisb) [reply completion/waffle.py:30]: My bet is that this is meant to be a platform-togglable flag, because I was bit by this in Tutor: https://discuss.overhang.io/t/course-not-starting/962/12 I assumed that the fact that the flag was not enabled by default was a feature, and not a bug. Also, this flag has been there for a very long time, so even if the flag was meant to be temporary, it's now permanent.

**Comment by regisb** [comment]
> I plan on delaying merge until I can see the monitoring custom attribute from the edx-platform PR. Please ensure this is squashed as you wish.

Yes, this makes sense. I rebased my changes so this should be ready to merge once you see the deprecation warnings in edx-platform.

File: completion/waffle.py
Comment (robrap) [reply completion/waffle.py:30]: Sounds good.  I'm not clear on why it is an option and why it is off by default.  I asked the team and we'll see.
Either way, we can merge and it can always be updated later.

**Pull request merged**
